### PR TITLE
Support avro timestamp-nanos logical type

### PR DESCRIPTION
**Pull request by grantatspothero**
## Description
Add support for `timestamp-nanos` to `NativeLogicalTypesAvroTypeBlockHandler`. 

Note this does not change the behavior of the hive connector, `HiveAvroTypeBlockHandler` still only supports `timestamp-millis`. `timestamp-micros` and `timestamp-nanos` logical types will continue to be ignored. 

## Additional context and related issues
Avro 1.12 spec added `timestamp-nanos` and `local-timestamp-nanos` logical types: https://avro.apache.org/docs/1.12.0/specification/ 

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

Change should not be visible by users, hive connector behavior is unchanged.

### CODE REVIEW

File: lib/trino-hive-formats/src/main/java/io/trino/hive/formats/avro/NativeLogicalTypesAvroTypeBlockHandler.java [hunk 1]
```diff
@@ -271,6 +282,27 @@ public void decodeIntoBlock(Decoder decoder, BlockBuilder builder)
...
+            longDecoder = requireNonNull(longRead, "longRead is null").getLongDecoder();
+        }
+
+        @Override
+        public void decodeIntoBlock(Decoder decoder, BlockBuilder builder)
+                throws IOException
+        {
+            long epochNanos = longDecoder.apply(decoder);
+            long epochMicros = epochNanos / NANOSECONDS_PER_MICROSECOND;
+            int picosOfMicro = (int) (epochNanos % NANOSECONDS_PER_MICROSECOND) * PICOSECONDS_PER_NANOSECOND;
```
Comment (grantatspothero) [inline lib/trino-hive-formats/src/main/java/io/trino/hive/formats/avro/NativeLogicalTypesAvroTypeBlockHandler.java]: note to self: add test for negative epoch values, % seems unsafe

File: lib/trino-hive-formats/src/main/java/io/trino/hive/formats/avro/NativeLogicalTypesAvroTypeManager.java [hunk 2]
```diff
@@ -216,6 +227,17 @@ private static BiFunction<Block, Integer, Object> getAvroFunction(AvroLogicalTyp
...
             }
+            case TimestampNanosLogicalType _ -> {
+                if (!(type instanceof TimestampType timestampType)) {
+                    throw new AvroTypeException("Can't represent Avro logical type %s with Trino Type %s".formatted(logicalType, type));
+                }
+                checkState(!timestampType.isShort(), "Short timestamp type cannot represent nanosecond precision");
+                yield (block, position) ->
+                {
+                    SqlTimestamp timestamp = ((SqlTimestamp) timestampType.getObjectValue(block, position));
+                    return (timestamp.getEpochMicros() * NANOSECONDS_PER_MICROSECOND) + (timestamp.getPicosOfMicros() / PICOSECONDS_PER_NANOSECOND);
```
Comment (findepi) [inline lib/trino-hive-formats/src/main/java/io/trino/hive/formats/avro/NativeLogicalTypesAvroTypeManager.java]: guard for overflow. 
add a test for 9999-12-31 timestamps (surprisingly popular in real world data)

File: lib/trino-hive-formats/src/test/java/io/trino/hive/formats/avro/TestAvroPageDataReaderWithAvroNativeTypeManagement.java [hunk 3]
```diff
@@ -265,37 +275,116 @@ public void testWriting()
...
+        try (AvroFileWriter fileWriter = new AvroFileWriter(
+                trinoLocalFilesystem.newOutputFile(testLocation).create(),
+                timestampsSchema,
+                new NativeLogicalTypesAvroTypeManager(),
+                AvroCompressionKind.NULL,
+                ImmutableMap.of(),
+                timestampsSchema.getFields().stream().map(Schema.Field::name).collect(toImmutableList()),
+                new NativeLogicalTypesAvroTypeBlockHandler().typeFor(timestampsSchema).getTypeParameters(), false)) {
+            fileWriter.write(timestampsPage);
+        }
```
Comment (findepi) [inline lib/trino-hive-formats/src/test/java/io/trino/hive/formats/avro/TestAvroPageDataReaderWithAvroNativeTypeManagement.java:348]: This can't be the only test. `NativeLogicalTypesAvroTypeManager` is our class and its logic may or may not be entirely correct.

There should be a test with hive writing the file. See example existing test https://github.com/trinodb/trino/blob/16b41f68e40b92fefce08fdc2b21b238f5e7f48b/testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestAvroSchemaLiteral.java#L52-L53

File: lib/trino-hive-formats/src/main/java/io/trino/hive/formats/avro/NativeLogicalTypesAvroTypeBlockHandler.java (hunk 1 above)
Comment (findepi) [reply lib/trino-hive-formats/src/main/java/io/trino/hive/formats/avro/NativeLogicalTypesAvroTypeBlockHandler.java]: it's not.
 use floorDiv and floorMod

File: lib/trino-hive-formats/src/main/java/io/trino/hive/formats/avro/NativeLogicalTypesAvroTypeManager.java (hunk 2 above)
Comment (grantatspothero) [reply lib/trino-hive-formats/src/main/java/io/trino/hive/formats/avro/NativeLogicalTypesAvroTypeManager.java]: guarded for overflow and added test for longest representable timestamp nanos.

File: lib/trino-hive-formats/src/test/java/io/trino/hive/formats/avro/TestAvroPageDataReaderWithAvroNativeTypeManagement.java (hunk 3 above)
Comment (grantatspothero) [reply lib/trino-hive-formats/src/test/java/io/trino/hive/formats/avro/TestAvroPageDataReaderWithAvroNativeTypeManagement.java:348]: added hive connector product test, note the test indicates no change in behavior for hive connector

File: lib/trino-hive-formats/src/main/java/io/trino/hive/formats/avro/NativeLogicalTypesAvroTypeBlockHandler.java [hunk 4]
```diff
@@ -271,6 +285,27 @@ public void decodeIntoBlock(Decoder decoder, BlockBuilder builder)
...
+        {
+            longDecoder = requireNonNull(longRead, "longRead is null").getLongDecoder();
+        }
+
+        @Override
+        public void decodeIntoBlock(Decoder decoder, BlockBuilder builder)
+                throws IOException
+        {
+            long epochNanos = longDecoder.apply(decoder);
+            long epochMicros = divideExact(epochNanos, NANOSECONDS_PER_MICROSECOND);
```
Comment (raunaqmorarka) [inline lib/trino-hive-formats/src/main/java/io/trino/hive/formats/avro/NativeLogicalTypesAvroTypeBlockHandler.java:303]: Why do we need `divideExact` ?
Usually we use `floorDiv`

Comment (findepi) [reply lib/trino-hive-formats/src/main/java/io/trino/hive/formats/avro/NativeLogicalTypesAvroTypeBlockHandler.java:303]: i suppose here we use divideExact not to be lossy (i don't know the context and don't know if being lossy is ok or not)
